WMFData/Sources/WMFData/Data Controllers/Donor Experience/WMFFundraisingCampaignDataController.swift

```diff
@@ -354,28 +355,76 @@ import Foundation
         
         return configs
     }
-    
-    private func randomAssetFrom(assets: [WMFFundraisingCampaignConfigResponse.FundraisingCampaignConfig.Asset]) -> WMFFundraisingCampaignConfigResponse.FundraisingCampaignConfig.Asset {
-        guard assets.count > 1 else {
-            return assets[0]
+
+
+    // MARK: - Deterministic bucketing
+
+    /// Stable hash mapped to [0, 1).
+    private func stableHash01(_ seed: String) -> Double {
+        // FNV-1a 64-bit
+        var hash: UInt64 = 1469598103934665603
+        let prime: UInt64 = 1099511628211
+        for b in seed.utf8 {
+            hash ^= UInt64(b)
+            hash &*= prime
         }
-        
-        // TODO: Is a seed needed?
-        let f = Float.random(in: 0..<1)
-            
-        var sum: Float = 0
-        for (i, asset) in assets.enumerated() {
-            
-            guard let weight = asset.weight else { continue }
-            
-            sum += weight
-            if f <= sum {
-                return assets[i]
+        return Double(hash) / Double(UInt64.max)
+    }
+
+    /// Deterministically picks one asset using (optionally) weighted A/B configuration.
+    /// - If all weights are nil → uniform distribution.
+    /// - If some weights are nil → they default to 1.0, then all weights are normalized.
+    private func randomAssetFrom(assets: [WMFFundraisingCampaignConfigResponse.FundraisingCampaignConfig.Asset], seed: String) -> WMFFundraisingCampaignConfigResponse.FundraisingCampaignConfig.Asset {
+        guard assets.count > 1 else { return assets[0] }
+
+        var weights: [Double] = []
+        weights.reserveCapacity(assets.count)
+
+        var sawNonNil = false
+        var sum: Double = 0
+
+        for a in assets {
+            if let w = a.weight {
+                sawNonNil = true
+                let v = max(0.0, Double(w)) // clamp negatives to 0
+                weights.append(v)
+                sum += v
+            } else {
+                weights.append(.nan) // mark as nil for now
             }
         }
-        
-        return assets[assets.count - 1]
+
+        // Default behavior for nil weights and normalization
+        if !sawNonNil {
+            let uniform = 1.0 / Double(assets.count)
+            weights = Array(repeating: uniform, count: assets.count)
+        } else {
+            // Treat nil weights as 1.0
+            for i in 0..<weights.count where weights[i].isNaN {
+                weights[i] = 1.0
+                sum += 1.0
+            }
+            if sum <= 0 {
+                let uniform = 1.0 / Double(assets.count)
+                weights = Array(repeating: uniform, count: assets.count)
+            } else {
+                for i in 0..<weights.count { weights[i] /= sum }
+            }
+        }
```

Comment:
All this normalization feels a little heavy to me. If we must ensure weight integrity (we can check with Android if they are), do you think we could do some simple weight additions and exit early with a nil asset if they don't add up to 1.0? Above on line 337 we can just guard against a nil randomAsset and `continue` in that case, similar to what we do when the value is empty.

The app wouldn't show a campaign, but I would rather fall back to that and force the config creator to write out the weights correctly than correct for them here.